### PR TITLE
Cleanup related to tasks and contexts

### PR DESCRIPTION
**Pull request by robhoes**
Especially note the changes to `Ref.t`. I have successfully run this through a BVT.

### CODE REVIEW

File: ocaml/xapi-types/ref.ml [hunk 1]
```diff
@@ -12,73 +12,63 @@
...
-let null = ref_prefix ^ "NULL"
+  Real uuid
 
-let string_of x = x
-
-let of_string x = x
-
-(* very ugly hack to be able to distinguish dummy and real tasks. Moreover, dummy tasks have their name *)
-(* embedded into the reference .... *)
+let null = Real "NULL"
```
[the remark this answers is not in the document]
Comment (robhoes) [reply ocaml/xapi-types/ref.ml]: I think making it a separate constructor is a good idea indeed.

File: ocaml/xapi-types/ref.ml [hunk 2]
```diff
@@ -12,73 +12,63 @@
...
+  | Dummy (uuid, name) -> dummy_prefix ^ dummy_sep ^ uuid ^ dummy_sep ^ name
+  | Other x -> x
 
-let name_of_dummy x =
-  match cut_in_three ~sep:dummy_sep x with
-  | Some(_, _, name) -> name
-  | None -> failwith (
-      Printf.sprintf "Ref.name_of_dummy: %s is not a valid dummy reference" x)
+let of_string x =
+  match Astring.String.cut ~sep:ref_prefix x with
```
[the remark this answers is not in the document]
Comment (robhoes) [reply ocaml/xapi-types/ref.ml]: I guess it does! It's a weird way of checking a prefix actually...

File: ocaml/xapi/context.ml [hunk 3]
```diff
@@ -59,7 +58,7 @@ let get_task_id ctx =
   ctx.task_id
 
 let task_in_database ctx =
-  ctx.task_in_database
+  not (Ref.is_dummy ctx.task_id)
```
[the remark this answers is not in the document]
Comment (robhoes) [reply ocaml/xapi/context.ml]: I don't think that we'd ever have a NULL ref in the context. But yes, we might as well add a `Ref.is_real` and use that.

**Comment by robhoes** [comment]
@edwintorok Added fixups for the three comments.

**Review by edwintorok — APPROVED** [review]
LGTM, but would be good for someone else to have a look as well.

File: ocaml/xapi-types/ref.ml [hunk 4]
```diff
@@ -12,73 +12,83 @@
...
-  Astring.String.is_prefix ~affix:dummy_prefix x
+let of_string x =
+  if x = ref_null then
+    Null
+  else
+    match Astring.String.cut ~sep:ref_prefix x with
+    | Some ("", uuid) -> Real uuid
+    | _ ->
+      match Astring.String.cuts ~sep:dummy_sep x with
+      | [prefix; uuid; name] when prefix = dummy_prefix -> Dummy (uuid, name)
```
Comment (jonludlam) [inline ocaml/xapi-types/ref.ml]:
```suggestion
      | prefix::uuid::name when prefix = dummy_prefix -> Dummy (uuid, String.concat dummy_sep name)
```

Comment (robhoes) [reply ocaml/xapi-types/ref.ml]: I've squashed this in.

**Pull request merged**